File: proper_cloud/scripts/nd_cloud.py
#!/usr/bin/env python
import tf
import rospy
import numpy as np

import sensor_msgs.point_cloud2 as pc2
from sensor_msgs.msg import PointCloud2


class NdCloud():
    def __init__(self):
        rospy.init_node('cloud')

        rospy.Subscriber(
            '/ueye/cloud', PointCloud2, self.cb_point_cloud, queue_size=5)

        self.pub_cloud = rospy.Publisher(
            '/ueye/scan', PointCloud2, queue_size=10)

        self.listener = tf.TransformListener()

        rospy.spin()

    def transform_point_cloud(self, matrix, cloud):
        points = np.hstack((cloud, np.ones((len(cloud), 1))))
        cloud = np.float32([np.dot(matrix, point)[:3] for point in points])
        return cloud

    def transform_matrix(self, target, source, stamp):
        self.listener.waitForTransform(
            target, source, stamp, rospy.Duration(1.0))
        (position, quaternion) = self.listener.lookupTransform(
            target, source, stamp)
        matrix = tf.transformations.quaternion_matrix(quaternion)
        matrix[:3, 3] = position
        return matrix

    def cb_point_cloud(self, msg_cloud):
        try:
            stamp = msg_cloud.header.stamp
            points = pc2.read_points(msg_cloud, skip_nans=False)
            points3d = np.float32([point for point in points])
            matrix1 = self.transform_matrix('/world', '/camera0', stamp)
            matrix2 = self.transform_matrix('/world', '/workobject', stamp)
            matrix = np.dot(np.linalg.inv(matrix2), matrix1)
            points3d = self.transform_point_cloud(matrix, points3d)
            # rospy.loginfo(points3d)
            cloud_out = PointCloud2()
            cloud_out.header.stamp = stamp
            cloud_out.header.frame_id = "/workobject"
            cloud_out = pc2.create_cloud_xyz32(cloud_out.header, points3d)
            self.pub_cloud.publish(cloud_out)
        except:
            pass


if __name__ == '__main__':
    try:
        NdCloud()
    except rospy.ROSInterruptException:
        pass

File: proper_cloud/src/nd_cloud.cpp
#include <ros/ros.h>

#include <sensor_msgs/PointCloud2.h>

#include <pcl_conversions/pcl_conversions.h>

#include <pcl/point_cloud.h>
#include <pcl/point_types.h>

#include <pcl_ros/point_cloud.h>
#include <pcl_ros/transforms.h>

#include <tf/transform_listener.h>


class NdCloud {
private:
  ros::NodeHandle nh;
  ros::Subscriber sub_cloud;
  ros::Publisher pub_cloud;

  tf::TransformListener *tf_listener;
  tf::StampedTransform transform;

  Eigen::Matrix4f matrix;
  Eigen::Matrix4f matrix1;
  Eigen::Matrix4f matrix2;
public:
  typedef pcl::PointXYZ Point;
  typedef pcl::PointCloud<Point> PointCloud;

  NdCloud() {
    sub_cloud = nh.subscribe<sensor_msgs::PointCloud2>("/ueye/cloud", 5,  &NdCloud::cbPointCloud, this);

    pub_cloud = nh.advertise<sensor_msgs::PointCloud2>("/ueye/scan", 10);

    tf_listener = new tf::TransformListener();
  }

  ~NdCloud() {}

  void cbPointCloud(const sensor_msgs::PointCloud2ConstPtr& cloud_msg) {
    // PointCloud cloud;
    // pcl::fromROSMsg(*cloud_msg, cloud);

    try {
      sensor_msgs::PointCloud2 cloud_out;

      // ROS_INFO_STREAM("Frame_id: " << (*cloud_msg).header.frame_id);
      // ROS_INFO_STREAM("Stamp: " << (*cloud_msg).header.stamp);
      // ROS_INFO_STREAM("Time: " << ros::Time::now());

      ros::Time stamp = (*cloud_msg).header.stamp;
      tf_listener->waitForTransform("/world", "/camera0", stamp, ros::Duration(1.0));
      tf_listener->lookupTransform("/world", "/camera0", stamp, transform);
      pcl_ros::transformAsMatrix(transform, matrix1);
      tf_listener->waitForTransform("/world", "/workobject", stamp, ros::Duration(1.0));
      tf_listener->lookupTransform("/world", "/workobject", stamp, transform);
      pcl_ros::transformAsMatrix(transform, matrix2);
      matrix = matrix2.inverse() * matrix1;
      pcl_ros::transformPointCloud(matrix, *cloud_msg, cloud_out);
      ROS_INFO_STREAM("Transform:" << matrix);

      cloud_out.header.frame_id = "/workobject";
      pub_cloud.publish(cloud_out);
    } catch (tf::TransformException ex) {
      ROS_ERROR("%s",ex.what());
    }

    // sensor_msgs::PointCloud2 output;
    // pcl::toROSMsg(cloud, output);
    // pub_cloud.publish(output);
  }
};


int main (int argc, char** argv) {
  ros::init(argc, argv, "cloud");

  NdCloud nd_cloud;

  ros::spin();
}
